### PR TITLE
Reduced number of transactions within the outbox executor

### PR DESCRIPTION
**Pull request by zaychenko-sergei**
## Description

Outbox main loop was revised to minimize the number of transactions:
 - split outbox into _planner_ and _consumption jobs_
 - planner analyzes current state and loads bunch of unprocessed messages within a 1 transaction ony
 - consumption jobs invoke consumers and detect their failures

## Checklist before requesting a review

- [x] Unit and integration tests added
  
- [x] Compatibility:
    
  - [x] Network APIs: ✅
    
  - [x] Workspace layout and metadata: ✅ 
    
  - [x] Configuration: ✅
    
  - [ ] Container images: ✅
- [ ] Documentation:
    
  - [x] [Changelog](./CHANGELOG.md): ✅
    
  - [ ] [Public documentation](https://github.com/kamu-data/kamu-docs/): ✅
  
- [ ] Downstream effects:
    
  - [ ] [kamu-node](https://github.com/kamu-data/kamu-node): ✅

### CODE REVIEW

File: src/utils/messaging-outbox/src/executors/outbox_consumption_iteration_planner.rs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+                .and_modify(|by_consumer: &mut HashMap<String, OutboxMessageID>| {
+                    by_consumer.insert(
+                        boundary.consumer_name.clone(),
+                        boundary.last_consumed_message_id,
+                    );
+                })
+                .or_insert_with(|| {
+                    HashMap::from([(boundary.consumer_name, boundary.last_consumed_message_id)])
+                });
+        }
```
Comment (sergiimk) [inline src/utils/messaging-outbox/src/executors/outbox_consumption_iteration_planner.rs:138]: So you basically have a list of tuples `(P, C, ID)` and you are doing a `group by P`. I think [`itertools::chunk_by`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.chunk_by) combined with the ability to `.collect()` an iterator of tuples into `HashMap` could reduce the amount of code here, something like:
```
let boundaries: Vec<_> = stream.collect();
boundaries.sort_by(|b| b.producer);
boundaries
.into_iter()
.chunk_by(|b| &b.producer)
.map(|(producer, bounds)| (
  producer, 
  bounds.map(|b| (b.consumer, b.last_id)).collect()
))

File: src/utils/messaging-outbox/src/executors/outbox_consumption_iteration_planner.rs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        // Group messages by producers
+        let mut unprocessed_messages_by_producer: HashMap<String, Vec<OutboxMessage>> =
+            HashMap::new();
+        for unprocessed_message in unprocessed_messages {
+            if let Some(entry) =
+                unprocessed_messages_by_producer.get_mut(&unprocessed_message.producer_name)
+            {
+                entry.push(unprocessed_message);
+            } else {
+                unprocessed_messages_by_producer.insert(
```
Comment (sergiimk) [inline src/utils/messaging-outbox/src/executors/outbox_consumption_iteration_planner.rs:260]: same here, probably can be simplified

**Review by sergiimk — APPROVED** [review]
Great improvement, only left a suggestion how to simplify some code

File: src/utils/messaging-outbox/src/executors/outbox_consumption_iteration_planner.rs (hunk 1 above)
Comment (zaychenko-sergei) [reply src/utils/messaging-outbox/src/executors/outbox_consumption_iteration_planner.rs:138]: The code certainly got Rustier, but the resulting "simplicity" is not convincing :-)

File: src/utils/messaging-outbox/src/executors/outbox_consumption_iteration_planner.rs (hunk 2 above)
Comment (zaychenko-sergei) [reply src/utils/messaging-outbox/src/executors/outbox_consumption_iteration_planner.rs:260]: Rustified

File: src/utils/messaging-outbox/src/executors/outbox_consumption_iteration_planner.rs (hunk 1 above)
Comment (zaychenko-sergei) [reply src/utils/messaging-outbox/src/executors/outbox_consumption_iteration_planner.rs:138]: Before I had to clone the consumer names, but now `chunk_by` forces to clone the producer name. It's probably irrelevant, we are talking about <10 records for now.

**Pull request merged**